File: code/online_challenges/src/leetcode/range_sum_of_bst/range_sum_of_bst.cpp
/* Part of Cosmos by OpenGenus Foundation */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
   
public:
    int gsum=0;
    int rangeSumBST(TreeNode* root, int L, int R) 
    {
        if(root == NULL) return gsum;
        stack<TreeNode*> s;
        s.push(root);
        while(!s.empty())
        {
            root = s.top();
            s.pop();
            if(root != NULL)
            {
                if(root -> val >= L && root -> val <= R) gsum += root -> val;
                if(root -> val > L) s.push(root -> left);
                if(root -> val < R) s.push(root -> right);
            }
        }
        
        return gsum;
        
    }
};

File: code/unclassified/src/range_sum_of_BST/range_sum_of_bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        if(root == null) return 0;
        int sum = 0;
        if(root.val >= low){
            sum+=rangeSumBST(root.left,low,high);
        } if(root.val >= low && root.val <= high){
            sum+=root.val;
        } if(root.val <= high){
            sum+=rangeSumBST(root.right,low,high);
        }
        return sum;
    }
}
